### PR TITLE
fix(linux): Fix installation of shared packages

### LINKED ISSUE #5311: bug(linux): packages installed shared do not appear in ibus
**Describe the bug**

A package installed to shared with does not appear in ibus, even after restarting.

```shell
km-package-install -s -f sample.kmp
```

---

**Keyman for Windows/macOS/Linux:**
 - OS: Ubuntu 20.04 
 - Keyman version: 14.02.76

### PR DESCRIPTION
**Pull request by ermshiperete**
This change fixes the installation of shared packages, so that they show up in the list of the installed keyboards.

Fixes #5311.

# User Testing

### SUITE_INSTALL_PKG: Install a package

- Install a keyboard to the shared area using the command line (e.g. `sudo km-package-install -s -p isis_bangla`). Check that the new keyboard appears in the dropdown with installed keyboards:

Screenshot from 2021-12-07 09-41-51

**TEST_INSTALL_PKG_GNOME:** Use a Linux version that uses Gnome (e.g. Ubuntu).

**TEST_INSTALL_PKG_IBUS:** same test on Wasta Linux

### SUITE_UNINSTALL: Uninstall a package

- Uninstall the keyboard installed above with `sudo km-package-uninstall -s isis_bangla`. This should remove the keyboard from the dropdown.

**TEST_UNINSTALL_PKG_GNOME:**  Use a Linux version that uses Gnome (e.g. Ubuntu).

**TEST_UNINSTALL_PKG_IBUS:** same test on Wasta Linux

### SUITE_INSTALL_KMP: Install a .kmp file

- Similar to the previous tests, but with a .kmp file already downloaded on the local machine: e.g. `sudo km-package-install -s -f khmer_angkor.kmp`

**TEST_INSTALL_KMP_GNOME:** Use a Linux version that uses Gnome (e.g. Ubuntu).

**TEST_INSTALL_KMP_IBUS:** same test on Wasta Linux

### CODE REVIEW

File: linux/keyman-config/keyman_config/gnome_keyboards_util.py [hunk 1]
```diff
@@ -1,23 +1,52 @@
...
 
     def read_input_sources(self):
-        sourcesVal = self.input_sources.get_value("sources")
-        sources = self._convert_variant_to_array(sourcesVal)
+        if self.is_sudo:
+            process = subprocess.run(
+                ['sudo', '-H', '-u', os.environ.get('SUDO_USER'),
+                 'DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/%s/bus' % os.environ.get('SUDO_UID'),
+                 'gsettings', 'get', 'org.gnome.desktop.input-sources', 'sources'],
+                capture_output=True)
```
Comment (mcdurdin) [inline linux/keyman-config/keyman_config/gnome_keyboards_util.py:27]: This pattern is repeated once in this file and also twice in ibus_util.py. Could we refactor these into a single function?

In fact, the entire wrapper around Gio.Settings could possibly be factored to a single helper that works in sudo and normal user mode?

File: linux/keyman-config/keyman_config/ibus_util.py [hunk 2]
```diff
@@ -10,6 +10,45 @@
...
+    def write_preload_engines(self, bus, preload_engines):
+        if os.environ.get('SUDO_USER'):
+            sourcesVal = str(preload_engines)
+            subprocess.run(
+                ['sudo', '-H', '-u', os.environ.get('SUDO_USER'),
+                 'DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/%s/bus' % os.environ.get('SUDO_UID'),
+                 'gsettings', 'set', 'org.freedesktop.ibus.general', 'preload-engines', sourcesVal])
+        else:
+            self.ibus_settings.set_strv("preload-engines", preload_engines)
+            bus.preload_engines(preload_engines)
```
Comment (mcdurdin) [inline linux/keyman-config/keyman_config/ibus_util.py:49]: This doesn't seem to be run in sudo mode?

**Review by mcdurdin — CHANGES_REQUESTED** [review]
Looks good. Only minor tidyup and one possible gap?

File: linux/keyman-config/keyman_config/gnome_keyboards_util.py [hunk 3]
```diff
@@ -1,81 +1,22 @@
...
     def __init__(self):
-        if os.environ.get('SUDO_USER'):
-            self.input_sources = None
-            self.is_sudo = True
-            logging.debug('Running with sudo. Real user: %s', os.environ.get('SUDO_USER'))
-        else:
-            self.input_sources = Gio.Settings.new("org.gnome.desktop.input-sources")
-            self.is_sudo = False
-            logging.debug('Running as regular user')
+        self.input_sources = GSettings('org.gnome.desktop.input-sources')
```
Comment (mcdurdin) [inline linux/keyman-config/keyman_config/gnome_keyboards_util.py:10]: 😍 So pretty! ✨

File: linux/keyman-config/keyman_config/gsettings.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
+#!/usr/bin/python3
+import logging
+import os
+import subprocess
+from gi.repository import Gio
+
+from gi.overrides.GLib import Variant
+
+
+class GSettings():
```
Comment (mcdurdin) [inline linux/keyman-config/keyman_config/gsettings.py:10]: This is lovely -- could you just add a comment on the front describing usage for Mr Posterity?

**Review by mcdurdin — COMMENTED** [review]
Changes look great! Just need to fix some tests e.g. `test_gnome_keyboards_util.GnomeKeyboardsUtilTests.test_ConvertArrayToVariantToArray_Empty` I think :grin:

File: linux/keyman-config/keyman_config/gsettings.py (hunk 4 above)
Comment (ermshiperete) [reply linux/keyman-config/keyman_config/gsettings.py:10]: done

**Review by ermshiperete — COMMENTED** [review]
Tests fixed :smiley_cat:

**Comment by MakaraSok** [comment]
## SUITE_INSTALL_PKG: Install a package

The version (15.0.166-alpha) of Keyman install from this PR went through without any err in Ubuntu, but there are some errs in wasta, i.e.

```
keyman@keyman-VirtualBox:~/Downloads/6015/archive-4/results$ sudo apt update
Hit:1 http://packages.sil.org/ubuntu focal InRelease
Hit:2 http://archive.ubuntu.com/ubuntu focal InRelease                         
Hit:3 http://ppa.launchpad.net/keymanapp/keyman/ubuntu focal InRelease         
Hit:4 http://archive.canonical.com/ubuntu focal InRelease                      
Hit:5 http://archive.ubuntu.com/ubuntu focal-updates InRelease                 
Get:6 http://archive.ubuntu.com/ubuntu focal-backports InRelease [108 kB]      
Hit:7 http://ppa.launchpad.net/wasta-linux/cinnamon-4-8/ubuntu focal InRelease 
Hit:8 http://ppa.launchpad.net/wasta-linux/wasta-apps/ubuntu focal InRelease   
Get:9 http://archive.ubuntu.com/ubuntu focal-security InRelease [114 kB]       
Hit:10 http://ppa.launchpad.net/wasta-linux/wasta/ubuntu focal InRelease       
Err:11 https://repo.skype.com/deb stable InRelease                             
  503  Service Unavailable [IP: 184.50.29.51 443]
Fetched 222 kB in 3s (71.5 kB/s)
Reading package lists... Done
Building dependency tree       
Reading state information... Done
416 packages can be upgraded. Run 'apt list --upgradable' to see them.
W: Failed to fetch https://repo.skype.com/deb/dists/stable/InRelease  503  Service Unavailable [IP: 184.50.29.51 443]
W: Some index files failed to download. They have been ignored, or old ones used instead.
keyman@keyman-VirtualBox:~/Downloads/6015/archive-4/results$ sudo dpkg -i \
> {ibus-keyman,ibus-kmfl,keyman,kmflcomp,libkmfl,libkmnkbp0,python3-keyman-config}*$(lsb_release -s -c)*.deb
(Reading database ... 432616 files and directories currently installed.)
Preparing to unpack ibus-keyman_15.0.166-1~PR-6015+focal11_amd64.deb ...
Unpacking ibus-keyman (15.0.166-1~PR-6015+focal11) over (13.0.104-1~sil2~focal) ...
Preparing to unpack ibus-kmfl_15.0.166-1~PR-6015+focal11_amd64.deb ...
Unpacking ibus-kmfl (15.0.166-1~PR-6015+focal11) over (13.0.104-1~sil2~focal) ...
Preparing to unpack keyman_15.0.166-1~PR-6015+focal11_all.deb ...
Unpacking keyman (15.0.166-1~PR-6015+focal11) over (13.0.104-1~sil4~focal) ...
Preparing to unpack kmflcomp_15.0.166-1~PR-6015+focal11_amd64.deb ...
Unpacking kmflcomp (15.0.166-1~PR-6015+focal11) over (13.0.104-1~sil2~focal) ...
Preparing to unpack libkmfl0_15.0.166-1~PR-6015+focal11_amd64.deb ...
Unpacking libkmfl0:amd64 (15.0.166-1~PR-6015+focal11) over (13.0.104-1~sil2~focal) ...
Preparing to unpack libkmflcomp0_15.0.166-1~PR-6015+focal11_amd64.deb ...
Unpacking libkmflcomp0:amd64 (15.0.166-1~PR-6015+focal11) over (13.0.104-1~sil2~focal) ...
Selecting previously unselected package libkmflcomp-dev:amd64.
Preparing to unpack libkmflcomp-dev_15.0.166-1~PR-6015+focal11_amd64.deb ...
Unpacking libkmflcomp-dev:amd64 (15.0.166-1~PR-6015+focal11) ...
Selecting previously unselected package libkmfl-dev:amd64.
Preparing to unpack libkmfl-dev_15.0.166-1~PR-6015+focal11_amd64.deb ...
Unpacking libkmfl-dev:amd64 (15.0.166-1~PR-6015+focal11) ...
Preparing to unpack libkmnkbp0-0_15.0.166-1~PR-6015+focal11_amd64.deb ...
Unpacking libkmnkbp0-0:amd64 (15.0.166-1~PR-6015+focal11) over (13.0.104-1~sil2~focal) ...
Preparing to unpack python3-keyman-config_15.0.166-1~PR-6015+focal11_all.deb ...
Unpacking python3-keyman-config (15.0.166-1~PR-6015+focal11) over (13.0.104-1~sil4~focal) ...
Setting up libkmflcomp0:amd64 (15.0.166-1~PR-6015+focal11) ...
Setting up libkmflcomp-dev:amd64 (15.0.166-1~PR-6015+focal11) ...
Setting up libkmnkbp0-0:amd64 (15.0.166-1~PR-6015+focal11) ...
dpkg: dependency problems prevent configuration of python3-keyman-config:
 python3-keyman-config depends on python3-sentry-sdk; however:
  Package python3-sentry-sdk is not installed.

dpkg: error processing package python3-keyman-config (--install):
 dependency problems - leaving unconfigured
Setting up ibus-keyman (15.0.166-1~PR-6015+focal11) ...
dpkg: dependency problems prevent configuration of keyman:
 keyman depends on python3-keyman-config; however:
  Package python3-keyman-config is not configured yet.

dpkg: error processing package keyman (--install):
 dependency problems - leaving unconfigured
Setting up kmflcomp (15.0.166-1~PR-6015+focal11) ...
Setting up libkmfl0:amd64 (15.0.166-1~PR-6015+focal11) ...
Setting up libkmfl-dev:amd64 (15.0.166-1~PR-6015+focal11) ...
Setting up ibus-kmfl (15.0.166-1~PR-6015+focal11) ...
Processing triggers for man-db (2.9.1-1) ...
Processing triggers for bamfdaemon (0.5.3+18.04.20180207.2-0ubuntu2) ...
Rebuilding /usr/share/applications/bamf-2.index...
Processing triggers for gnome-menus (3.36.0-1ubuntu1) ...
Processing triggers for desktop-file-utils (0.24-1ubuntu3) ...
Processing triggers for mime-support (3.64ubuntu1) ...
Processing triggers for libglib2.0-0:amd64 (2.64.6-1~ubuntu20.04.2) ...
Processing triggers for hicolor-icon-theme (0.17-2) ...
Processing triggers for shared-mime-info (1.15-1) ...
Processing triggers for libc-bin (2.31-0ubuntu9.2) ...
Errors were encountered while processing:
 python3-keyman-config
 keyman
keyman@keyman-VirtualBox:~/Downloads/6015/archive-4/results$ sudo apt install -f
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Correcting dependencies... Done
The following additional packages will be installed:
  python3-sentry-sdk
The following NEW packages will be installed:
  python3-sentry-sdk
0 upgraded, 1 newly installed, 0 to remove and 408 not upgraded.
2 not fully installed or removed.
Need to get 83.6 kB of archives.
After this operation, 477 kB of additional disk space will be used.
Do you want to continue? [Y/n] y
Get:1 http://ppa.launchpad.net/keymanapp/keyman/ubuntu focal/main amd64 python3-sentry-sdk all 1.1.0-3~focal [83.6 kB]
Fetched 83.6 kB in 1s (93.5 kB/s)             
Selecting previously unselected package python3-sentry-sdk.
(Reading database ... 432665 files and directories currently installed.)
Preparing to unpack .../python3-sentry-sdk_1.1.0-3~focal_all.deb ...
Unpacking python3-sentry-sdk (1.1.0-3~focal) ...
Setting up python3-sentry-sdk (1.1.0-3~focal) ...
Setting up python3-keyman-config (15.0.166-1~PR-6015+focal11) ...
Setting up keyman (15.0.166-1~PR-6015+focal11) ...
postinst called to compile glib schemas

```

- **TEST_INSTALL_PKG_GNOME: FAILED**
  An error occurs after running `km-package-install -s -p isis_bangla`:
  ```
  ERROR:km-package-install: error: Failed to install keyboard package isis_bangla
  ERROR:You do not have permissions to install the keyboard files to the shared area /usr/local/share/keyman
  ```
- **TEST_INSTALL_PKG_IBUS: FAILED** The same error as in Ubuntu.


## SUITE_UNINSTALL: Uninstall a package
- **TEST_UNINSTALL_PKG_GNOME: FAILED** There is an error. See the screenshot.
  image
  
- **TEST_UNINSTALL_PKG_IBUS: FAILED** The same error as in Ubuntu.


## SUITE_INSTALL_KMP: Install a .kmp file
- **TEST_INSTALL_KMP_GNOME: FAILED** There is an error after running the command. See the screenshot.6015s
  image
- **TEST_INSTALL_KMP_IBUS: FAILED** The same error as in Ubuntu.

**Comment by ermshiperete** [comment]
@MakaraSok ahh, very sorry, my instructions were incomplete! You'll have to use `sudo` to run those commands. I updated the instructions.
@keymanapp-test-bot retest

**Comment by MakaraSok** [comment]
## SUITE_INSTALL_PKG: Install a package

Ubuntu 20.04.3 LTS
Wasta Ubuntu 20.04 focal (x86-64), Cinnamon version 4.8.6

- **TEST_INSTALL_PKG_GNOME: FAILED** The keyboard seems to be successfully installed despite a few warnings after running the command to install it:
   ```
   ubuntu@ubuntu:~/Downloads/6015/archive-4/results$ sudo km-package-install -s -p isis_bangla
   Can't connect to IBus.
   WARNING:Unknown vkey: 106
   WARNING:Unknown vkey: 107
   WARNING:Unknown vkey: 109
   ```

  The ISIS (Blanga) keyboard is seen in the Keyman Configuration and Langauge Settings, but **not the dropdown**. See the screenshot below.
  
  image
  
- **TEST_INSTALL_PKG_IBUS: PASSED** Despite a few warnings after running the command, the keyboard does get installed and shown in the dropdown.

  image

  The warnings:
  ```
  keyman@keyman-VirtualBox:~$ sudo km-package-install -s -p isis_bangla
  [sudo] password for keyman: 
  WARNING:Unknown vkey: 106
  WARNING:Unknown vkey: 107
  WARNING:Unknown vkey: 109
  WARNING:could not find connected IBus.Bus
  ```


## SUITE_UNINSTALL: Uninstall a package
- **TEST_UNINSTALL_PKG_GNOME: PASSED?** The keyboard was removed from the Keyman Configuration and the Language Settings, but nothing is removed from the dropdown as it was not even there in the first place. :)

- **TEST_UNINSTALL_PKG_IBUS: PASSED** The keyboard was successsfully uninstalled and removed from the dropdown despite the following warning:

  ```
  keyman@keyman-VirtualBox:~$ sudo km-package-uninstall -s isis_bangla
  WARNING:could not find connected IBus.Bus
  ```


## SUITE_INSTALL_KMP: Install a .kmp file
- **TEST_INSTALL_KMP_GNOME: FAILED** There is an error after running the command read "Can't connect to IBUS". See the screenshot.

  image
  
- **TEST_INSTALL_KMP_IBUS: PASSED** The keyboard got installed successfully and does shown in the dropdown.

**Comment by MakaraSok** [comment]
After running the command, the result is shown below and the keyboard shown after rebooting.

```ubuntu@ubuntu:~/Downloads/6015/archive-4/results$ ps -ef | grep ibus-daemon
ubuntu      6599    5224  0 13:23 pts/0    00:00:00 grep --color=auto ibus-daemon
```

**Comment by MakaraSok** [comment]
## SUITE_INSTALL_PKG: Install a package
- **TEST_INSTALL_PKG_GNOME: PASSED** the keyboard does show up in the dropdown, but there are three warnings after running the command to install it.

```
sudo km-package-install -s -p isis_bangla
WARNING:Unknown vkey: 106
WARNING:Unknown vkey: 107
WARNING:Unknown vkey: 109
```

## SUITE_UNINSTALL: Uninstall a package
- **TEST_UNINSTALL_PKG_GNOME: PASSED** The keyboard is indeed got uninstalled and removed from the dropdown menu.

## SUITE_INSTALL_KMP: Install a .kmp file
- **TEST_INSTALL_KMP_GNOME: PASSED** the keyboard got installed and shown in the dropdown menu after running the command.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.174-alpha](https://keyman.com/downloads/releases/alpha/15.0.174)

**Pull request merged**